### ISSUE TITLE
minimist security issue

### ISSUE THREAD

**Issue by mgol** [foreversd/forever#1072]
GitHub reports a security advisory of all users of minimist `<1.2.2`:
https://github.com/advisories/GHSA-7fhm-mqm4-2wp7

forever is depending on the deprecated `optimist` package: https://www.npmjs.com/package/optimist. That package depends on minimist `~0.0.1`.

The solution is to switch to a recent `minimist` directly or to `yargs` or a similar package.

**Comment by yetzt** [comment]
the best solution would be @substack fixing the dependency in `optimist` by the press of one button https://github.com/substack/node-optimist/issues/152

**Comment by kibertoad** [comment]
I'll look into replacing optimist, as that library is dead

**Comment by kibertoad** [comment]
Replacing optimist will take some time that I don't currently have, unfortunately; if anyone would volunteer to make a PR, that would be appreciated greatly.

Meanwhile I've pinged @substack wrt to lending a hand to fix security issue on the optimist side, hopefully that works out as a stopgap solution.

**Comment by Ratstail91** [comment]
Hey, I replaced optimist with yargs, which seems like a decent drop-in replacement. Hopefully #1093 resolves this.

It passes all the tests, so it certainly should.

**Comment by hc-12** [comment]
@Ratstail91 it seems to be a partial fix as there are nested dependencies that still relies on minimist.

```json
    {
          "id": 1179,
          "path": "forever>flatiron>broadway>nconf>optimist>minimist",
          "dev": true,
          "optional": false,
          "bundled": false
        },
        {
          "id": 1179,
          "path": "forever>forever-monitor>broadway>nconf>optimist>minimist",
          "dev": true,
          "optional": false,
          "bundled": false
        },
        {
          "id": 1179,
          "path": "forever>flatiron>optimist>minimist",
          "dev": true,
          "optional": false,
          "bundled": false
        }
```

**Comment by kibertoad** [comment]
@hc-12 Aren't all of them devDependencies?

**Comment by hc-12** [comment]
> @hc-12 Aren't all of them devDependencies?

This is the audit report from npm with the latest `forever` version 3.0.1. It did remove an entry with latest patch which leaves us 3 remaining with minimist vulnerabilities.

**Comment by Ratstail91** [comment]
Ah crap. I guess it's up to me to fix these lol.

* [Flatiron](https://github.com/flatiron/flatiron) (a https framework?) Is currently out of date and ignored (the latest commit was 5 years ago)
* [Broadway](https://github.com/mbebenita/Broadway) (a video decoder?) is seemingly ignored too (with the last commit from 2 years ago)

Is there potentially a way to remove these dependencies entirely? What do they do?

Let's focus on flatiron first, since that will remove two of the issues.

**Comment by Ratstail91** [comment]
Actually, I'm sorry, but after going through the code, it seems forever heavily depends on flatiron. There's nothing I can do without rewriting it, or rewriting flatiron, and I don't get paid enough for that.

For those who come after, I wish you the best of luck.

**Comment by fool-derek** [comment]
Any progress for this issue?

**Comment by Ratstail91** [comment]
> Any progress for this issue?

No, sorry. This project us deprecated. I'd recommend using pm2.

**Comment by fool-derek** [comment]
> > Any progress for this issue?
> 
> No, sorry. This project us deprecated. I'd recommend using pm2.

Thanks for you quick answer!
